packages/core-bridge/src/helpers/abort_controller.rs

```diff
@@ -0,0 +1,206 @@
+use std::sync::{
+    Arc, OnceLock,
+    atomic::{AtomicU8, Ordering, fence},
+};
+
+use neon::{
+    handle::{Handle, Root},
+    object::Object,
+    prelude::{Context, JsResult},
+    types::{JsFunction, JsObject, JsValue},
+};
+
+use super::{BridgeResult, JsCallback, TryIntoJs, errors::IntoThrow as _};
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+pub type JsAbortController = JsObject;
+pub type JsAbortSignal = JsValue;
+
+/// An object that models a JavaScript `AbortController`, and its corresponding `AbortSignal`,
+/// allowing the Rust side to fire that signal if/when needed, e.g. when dropped from the Rust side.
+///
+/// The JS counterpart object is lazily instantiated when the signal gets converted to JS (through
+/// the `TryIntoJs` trait); this ensures that the Rust side can be created without a JS `Context`.
+pub struct AbortController {
+    inner: Arc<AbortControllerInner>,
+    drop_abort_reason: String,
+}
+
+/// An object that models the signal of a JavaScript `AbortController`.
+pub struct AbortSignal {
+    inner: Arc<AbortControllerInner>,
+}
+
+/// The inner state of an `AbortController`, shared between the Rust and JS sides.
+struct AbortControllerInner {
+    // The fact that we require a `Context` to initialize the JS counterpart means that we are running
+    // on the Node's thread, which guarantees that there can't be multiple threads calling into that
+    // function concurrently; that should in theory aleviate the need to use a lock on `js_counterpart`.
+    //
+    // It is however possible for the rust-side controller to get aborted from a non-Node thread
+    // while the JS-side controller is being created on the Node thread, in which case we don't
+    // want the Rust-side thread to get blocked for the JS-side to complete instantiation.
+    //
+    // By modelling the "JS initialization" and "is aborted" states as two distinct independant
+    // structures, we ensure that we're never blocking execution of either thread. This however
+    // means that either step may happen before the other, so we need to be careful not to miss
+    // sending the abort signal. The good news is that nothing bad will happen if we call the JS
+    // abort callback multiple times.
+    js_counterpart: OnceLock<AbortControllerJsCounterpart>,
+    aborted: OnceLock<String>,
+
+    state: AtomicU8,
+}
+
+struct AbortControllerJsCounterpart {
+    controller: Root<JsObject>,
+    abort: JsCallback<(String,), ()>,
+}
+
+const STATE_UNINITIALIZED: u8 = 0;
+const STATE_ARMED: u8 = 1;
+const STATE_ABORTED: u8 = 2;
```

Comment:
I don't see where this state is ever set?

Reply:
They are used in all operations on the atomic `state` variable, below.

Reply:
Sorry, that wasn't clear. I meant to highlight specifically ABORTED - I don't see where it ever gets set?

Reply:
Lines 157 and 185...
And ABORTED is at line 202.